engine/src/exec-engine.cpp

```diff
@@ -1369,40 +1369,73 @@ static void MCEngineSendOrCall(MCExecContext& ctxt, MCStringRef p_script, MCObje
 	}
 	if ((stat = optr->message(*t_message, t_params, p_is_send, True)) == ES_NOT_HANDLED)
 	{
+        // The message was not handled by the target object, so this is
+        // just a bunch of script to be executed as if it were in a handler
+        // in the target object (using domess).
+        
 		MCHandler *t_handler;
 		t_handler = optr -> findhandler(HT_MESSAGE, *t_message);
 		if (t_handler != NULL && t_handler -> isprivate())
+        {
 			ctxt . LegacyThrow(EE_SEND_BADEXP, *t_message);
-		else
-		{
-            MCAutoStringRef tptr;
+            goto cleanup;
+        }
 
-			if (t_params != NULL)
-			{
-                MCAutoValueRef t_value;
-				/* UNCHECKED */ t_params->eval(ctxt, &t_value);
-                MCAutoStringRef t_value_string;
-				ctxt . ConvertToString(*t_value, &t_value_string);
-                MCStringFormat(&tptr, "%@ %@", *t_message, *t_value_string);
-				
-			}
-            else
-                tptr = MCNameGetString(*t_message);
+        // The 'split into message and parameters' function above is used
+        // to ensure that all the parameters are evaluated in the current
+        // context (not the target). Since domess just takes a string, we
+        // convert the entire script back into a string with the params
+        // having been evaluated. This means in particular that variables
+        // containing arrays will not work here - they will be converted to
+        // the empty string.
+        
+        MCAutoStringRef tptr;
+        MCAutoListRef t_param_list;
+        MCListCreateMutable(',', &t_param_list);
+        MCParameter *t_param_ptr;
+        t_param_ptr = t_params;
+        
+        bool t_has_params;
+        t_has_params = t_params != nil;
+        while (t_param_ptr != NULL)
+        {
+            MCAutoValueRef t_value;
+            MCAutoStringRef t_value_string;
             
-			if ((stat = optr->domess(*tptr)) == ES_ERROR)
-				ctxt . LegacyThrow(EE_STATEMENT_BADCOMMAND, *t_message);
-		}
+            if (!t_param_ptr->eval(ctxt, &t_value) ||
+                !ctxt . ConvertToString(*t_value, &t_value_string) ||
+                !MCListAppend(*t_param_list, *t_value_string))
+                goto cleanup;
+
+            t_param_ptr = t_param_ptr -> getnext();
+        }
+        
+        if (t_has_params)
+        {
+            MCAutoStringRef t_params_string;
+            if (!MCListCopyAsString(*t_param_list, &t_params_string) ||
+                !MCStringFormat(&tptr, "%@ %@", *t_message, *t_params_string))
```

Comment:
Exactly the same in terms of allocations. Actually the best thing to do would be to change
`MCListCopyAsStringAndRelease` to make it give the buffer to the the string.